libp2p/src/builder.rs

```diff
@@ -0,0 +1,282 @@
+use libp2p_core::{muxing::StreamMuxerBox, Transport};
+use std::marker::PhantomData;
+
+pub struct TransportBuilder {
+    keypair: libp2p_identity::Keypair,
+}
+
+impl TransportBuilder {
+    pub fn new(keypair: libp2p_identity::Keypair) -> Self {
+        Self { keypair }
+    }
+
+    #[cfg(feature = "async-std")]
+    pub fn with_async_std(self) -> TcpBuilder<AsyncStd> {
+        TcpBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+        }
+    }
+
+    #[cfg(feature = "tokio")]
+    pub fn with_tokio(self) -> TcpBuilder<Tokio> {
+        TcpBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+        }
+    }
+}
+
+pub struct TcpBuilder<P> {
+    keypair: libp2p_identity::Keypair,
+    phantom: PhantomData<P>,
+}
+
+#[cfg(all(feature = "async-std", feature = "tcp"))]
+impl TcpBuilder<AsyncStd> {
+    pub fn with_tcp(self) -> RelayBuilder<AsyncStd, impl BuildableTransport> {
+        RelayBuilder {
+            transport: libp2p_tcp::async_io::Transport::new(
+                libp2p_tcp::Config::new().nodelay(true),
+            )
+            .upgrade(libp2p_core::upgrade::Version::V1)
+            .authenticate(libp2p_noise::Config::new(&self.keypair).unwrap())
+            .multiplex(libp2p_yamux::Config::default())
+            .map(|(p, c), _| (p, StreamMuxerBox::new(c))),
+            keypair: self.keypair,
+            phantom: PhantomData,
+        }
+    }
+}
+
+#[cfg(all(feature = "tokio", feature = "tcp"))]
+impl TcpBuilder<Tokio> {
+    pub fn with_tcp(self) -> RelayBuilder<Tokio, impl BuildableTransport> {
+        RelayBuilder {
+            transport: libp2p_tcp::tokio::Transport::new(libp2p_tcp::Config::new().nodelay(true))
+                .upgrade(libp2p_core::upgrade::Version::V1)
+                .authenticate(libp2p_noise::Config::new(&self.keypair).unwrap())
+                .multiplex(libp2p_yamux::Config::default())
+                .map(|(p, c), _| (p, StreamMuxerBox::new(c))),
+            keypair: self.keypair,
+            phantom: PhantomData,
+        }
+    }
+}
+
+// TODO: without_tcp
+
+pub struct RelayBuilder<P, T> {
+    transport: T,
+    keypair: libp2p_identity::Keypair,
+    phantom: PhantomData<P>,
+}
+
+#[cfg(feature = "relay")]
+impl<P, T: BuildableTransport> RelayBuilder<P, T> {
+    pub fn with_relay(
+        self,
+        relay_transport: libp2p_relay::client::Transport,
+    ) -> OtherTransportBuilder<P, impl BuildableTransport> {
```

Comment:
Once we make a `SwarmBuilder` out of this, it would be great if calling `with_relay` just changes the closure for `with_behaviour` from `(identity::Keypair) -> NB` to `(identity::Keypair, relay::Behaviour) -> NB`.